File: 1000cats.c
#include <stdio.h>
#include <time.h>
#include <stdlib.h>
#include <string.h>

void randstr(char *, int, char *);
void mutate(char *, char *, int);
int getscore(char *, char *);

int main(void)
{
  char target[] = "METHINKS IT IS LIKE A WEASEL";
  char characters[] = " ABCDEFGHIJKLMNOPQRSTUVWXYZ";
  int target_length = strlen(target);
  char survivor[target_length]; 
  int generation = 1;
  int offspring_count = 1000;
  int chance = 2;
  int max_score;
  int max_ordinal;
  char max_offspring[target_length + 1];
  int target_score = target_length;
  int c;
  int offspring_score;

  srand(time(NULL));
  memset(survivor, '\0', sizeof(survivor));
  memset(max_offspring, '\0', sizeof(max_offspring));
  randstr(survivor, target_length, characters);
  max_score = getscore(survivor, target);
  strcpy(max_offspring, survivor);
  printf("0: %s -- score: %d\n", max_offspring, max_score);
  while (max_score < target_score)
  {
    for (c = 0; c < offspring_count; c += 1)
    {
      mutate(survivor, characters, chance);
      offspring_score = getscore(survivor, target);
      if (offspring_score > max_score)
      {
        max_score = offspring_score;
	strcpy(max_offspring, survivor);
      }
    }
    printf("%d: %s -- score: %d\n", generation, max_offspring, max_score);
    generation += 1;
    strcpy(survivor, max_offspring);
  }
}

/* 
 * randstr
 * generates a random string composed of the elements in the string pointed to by *characters
 * the string is stored in the array pointed to by *string
 * length is the lenth of the string to generate
 */
void randstr(char *string, int length, char *characters)
{
  int range = strlen(characters);
  int c;
  int n;

  for (c = 0; c < length; c += 1)
  {
    n = rand() % range;
    string[c] = characters[n];
  }
  string[length] = '\0';
}

/*
 * mutate
 * with a probability of chance% changes each charter in random to a character in characters
 */
void mutate(char *random, char *characters, int chance)
{
  int c;
  int n;
  int p;
  int x = strlen(characters);

  for (c = 0; c < strlen(random); c += 1)
  {
    n = rand() % 100;
    if (n <= chance)
    {
      p = rand() % x;
      random[c] = characters[p];
    }
  }
}

/*
 * getscore
 * returns the number of characters that are identical between string1 and string2
 * both strings should have the same length
 */
int getscore(char *string1, char *string2)
{
  int length1 = strlen(string1);
  int length2 = strlen(string2);
  int c;
  int score = 0;

  if (length1 != length2)
  {
    printf("strings are not same length:\n%s\n%s\n can't continue\n", string1, string2);
    exit(1);
  }
  for (c = 0; c < length1; c += 1)
  {
    if (string1[c] == string2[c])
      score += 1;
  }
  return(score);
}

File: 1000cats.pl
#!/usr/bin/perl

use warnings;
use strict;

use String::Random;

my $target = "METHINKS IT IS LIKE A WEASEL";
my $characters = new String::Random;
$characters->{'S'} = ['A'..'Z', ' '];
my $generation = 1;
my $offspring_count = 1000;
my $chance = 2;
my $max_score;
my $max_ordinal;
my $max_offspring;
my $target_score = length($target);
my $c;
my $offspring_score;

my $survivor = $characters->randpattern("SSSSSSSSSSSSSSSSSSSSSSSSSSSS");
$max_score = getscore($survivor, $target);
$max_offspring = $survivor;
print "0: $max_offspring -- score: $max_score\n";
while ($max_score < $target_score)
{
  for ($c = 0; $c < $offspring_count; $c++)
  {
    $survivor = mutate($survivor, $chance);
    $offspring_score = getscore($survivor, $target);
    if ($offspring_score > $max_score)
    {
      $max_score = $offspring_score;
      $max_offspring = $survivor;
    }
  }
  print "$generation: $max_offspring -- score: $max_score\n";
  $generation++;
  $survivor = $max_offspring;
}

sub mutate
{
  my ($random, $chance) = @_;

  my $c;
  my $n;
  my $p;
  my $ret = $random;

  for ($c = 0; $c < length($random); $c++)
  {
    $n = rand(100);
    substr( $ret, $c, 1, $characters->randpattern("S") ) if $n <= $chance;
  }
  return $ret;
}

sub getscore
{
  my ($string1, $string2) = @_;

  my $length1 = length($string1);
  my $length2 = length($string2);
  my $c;
  my $score = 0;

  die "strings are not same length:\n$string1\n$string2\n can't continue\n" unless ($length1 == $length2);
  
  for ($c = 0; $c < $length1; $c++)
  {
    $score++ if ( substr($string1, $c, 1) eq substr($string2, $c, 1) );
  }
  return $score;
}
